freedomofpress/securedrop
# Ensure compatibility between Python requirements files

## Status

Ready for review.

## Description of Changes

Fixes #1918. Fixes #1807. Progress towards #1409.

Changes proposed in this pull request:

Adds a new utility to ensure compatibility between our requirements file. An order of importance has been set, `securedrop > test > testinfra > develop`, to ensure that our app code dependencies are never held back in order to remain compatible with less important dependencies (instead, it's vice-versa). Whenever a change is made to a particular `.in` file, one of the new `make` commands I've introduced should be run according to the new developer documentation. For granularity, this command will only compile the file you've updated and the ones "downstream" of it if necessary, in order to maintain compatibility. Further, it detects if changes are only "cosmetic" and won't update ".txt" files if all that's changed is letter case or comments; it's mostly idempotent.

The technical documentation in `update_python_dependencies.py` is pretty good I think, and the strategy is pretty simple to understand. I've added further documentation for developers/ contributors in our developer docs. 

New testinfra tests confirm that this utility is working as intended.

Our existing Python dependencies were brought into compatibility using `make update-python-dependencies-testinfra` since there were no conflicts between the securedrop and test requirements, and I intended only to make the minimum necessary changes for compatibility.

Other changes:
* All Python packages pre-installed in Travis are uninstalled at the start of each run to ensure a clean slate.
* Ansible is unpegged in `develop-requirements.txt` as the #1146 is resolved.

I agree with @conorsch that it's better simply not to install dependencies side-by-side that belong on the same machine, but this provides a short-term solution as it may be some months before SecureDrop is able to implement and transition to the model he describes in #1918.

## Testing

How should the reviewer test this PR?

Test that the changes to the testinfra and develop requirements have not broken anything and in fact show that the compatibility tool is working as intended (the new development testinfra tests will confirm this, as will Travis passing):

1. Install the new development requirements: `workon securedrop && pip-sync securedrop/requirements/develop-requirements.txt`.
2. Provision twice from a clean VM: `vagrant destroy development -f && vagrant up && vagrant provision`.
3. Run the configuration tests: `./testinfra/test.py development`.
4. Repeat steps 2-3 for the staging environment.

Make various changes to `in` files, and follow the developer documentation instructions on updating the `txt` files. Ensure the results are sane, idempotent, and granular. Changes to `testinfra-requirements.in`, e.g., should never induce changes to `securedrop-requirements.txt` or `test-requirements.txt`, if you use the make command as instructed in the new developer docs.

## Deployment

Any special considerations for deployment? Consider both:

1. Upgrading existing production instances.
2. New installs.

I don't believe so.

## Checklist

### If you made changes to the system configuration:


### If you made changes to documentation:
